### PR TITLE
Add basic support and documentation for CMake build system

### PR DESCRIPTION
**Pull request by ParticleG**

### CODE REVIEW

**Comment by masatake** [comment]
Thank you. Could you tell me how to build libreadtags with your change?
I'm using Fedora. On the platform, cmake command is available.
I would like to add 'building libreadtags with cmake' to the test cases run on GitHub Actions.

**Comment by ParticleG** [comment]
Thank you for your reply. I only use the main source files (` readtag. c ` and ` readtag. h `) to build into the static library. I currently do not have a Linux development environment, but you can try building it using the following command:

``` bash
cd ${source_dir}
mkdir build_debug
cmake -DCMAKE_BUILD_TYPE=Debug -S . -B build_debug
cmake --build build_debug --target libreadtags -j ${cpu_thread_count}
```

**Comment by ParticleG** [comment]
I'm currently configuring my WSL environment, I have never used autoconf or automake before. Can you provide me with a generated makefile? If you can also provide me with a makefile for the test directory, I should also be able to complete the CMakeList.txt for the test directory.

File: README.md [hunk 1]
```diff
@@ -10,3 +11,20 @@ NEWS.md describes note worty changes.
...
+
+## Usage
+
+### CMake Integration
+```cmake
+include(FetchContent)
+
+FetchContent_Declare(
+        libreadtags
+        GIT_REPOSITORY https://github.com/ParticleG/libreadtags.git
```
Comment (k-takata) [inline README.md]: The URL should be updated when merging into this repository, correct?
```suggestion
        GIT_REPOSITORY https://github.com/universal-ctags/libreadtags.git
```

Comment (ParticleG) [reply README.md]: Yes, thank you for pointing out

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.25)
+project(libreadtags)
+
+set(CMAKE_C_STANDARD 11)
```
Comment (masatake) [inline CMakeLists.txt]: Unlike ctags itself, I would like to keep libreadtags compilable with C99.

**Comment by masatake** [comment]
Could you tell me the use case of the file?

If you assume cmake as an alternative build system to autotools<1>, I expect cmake to provide libreadtags.**so** with soname and an install target.

If you assume cmake is just for building the library at a specified directory<2>, maybe the current version you submitted here is enough for merging.

**Comment by ParticleG** [comment]
> Could you tell me the use case of the file?
> 
> If you assume cmake as an alternative build system to autotools<1>, I expect cmake to provide libreadtags.**so** with soname and an install target.
> 
> If you assume cmake is just for building the library at a specified directory<2>, maybe the current version you submitted here is enough for merging.

cmake could create an install target with `install()` [method](https://cmake.org/cmake/help/latest/command/install.html), and canbe executed with `cmake --build ${build_dir} --target install`. But I currently don't have the makefile generated by the autotools for CMake coding reference.

Current `CMakeList.txt` is enough for other developers to use libreadtags in their CMake project as a dependency library

**Comment by ParticleG** [comment]
I am not very familiar with the `autoconf+automake` system. Based on my search results, the vast majority of configuration parameters should be in `configure.ac` and `makefile.am`? If that's true, maybe I could try to write an install target

File: CMakeLists.txt (hunk 2 above)
Comment (ParticleG) [reply CMakeLists.txt]: I've just found `AC-PROG-CC-C99` in configure.ac. I'll update that.

**Comment by ParticleG** [comment]
I'm using the command in the `.circleci` folder to configure the project. I see the `LT_VERSION` is [2:2:1] which would result in `libreadtags.so.1.1.2`, but the project version is 0.3.0 in `AC_INIT` step. May I ask why they are not the same?

**Comment by ParticleG** [comment]
I've updated `README.md` for how to build & install a shared lib. If you follow these commands, it should generate `libreadtags.so` and corresponding sonames (Current are `libreadtags.so.1` and `libreadtags.so.1.1.2`), and install to the `/usr/local` directory.

If there's anything still missing (like installing man or other stuff), just tell me to add them

**Comment by masatake** [comment]
> I'm using the command in the .circleci folder to configure the project. I see the LT_VERSION is [2:2:1] which would result in libreadtags.so.1.1.2, but the project version is 0.3.0 in AC_INIT step. May I ask why they are not the same?

LT_VERSION specifies the version of the library interface.
So even if we add a commit, we will not update the LT_VERSION as far as the interface is not changed. e.g. fixing a bug in a test case.

Could you give me time? I must understand your changes well.

**Comment by ParticleG** [comment]
> > I'm using the command in the .circleci folder to configure the project. I see the LT_VERSION is [2:2:1] which would result in libreadtags.so.1.1.2, but the project version is 0.3.0 in AC_INIT step. May I ask why they are not the same?
> 
> LT_VERSION specifies the version of the library interface. So even if we add a commit, we will not update the LT_VERSION as far as the interface is not changed. e.g. fixing a bug in a test case.
> 
> Could you give me time? I must understand your changes well.

Of course

**Comment by ParticleG** [comment]
> > I'm using the command in the .circleci folder to configure the project. I see the LT_VERSION is [2:2:1] which would result in libreadtags.so.1.1.2, but the project version is 0.3.0 in AC_INIT step. May I ask why they are not the same?
> 
> LT_VERSION specifies the version of the library interface. So even if we add a commit, we will not update the LT_VERSION as far as the interface is not changed. e.g. fixing a bug in a test case.
> 
> Could you give me time? I must understand your changes well.

I forgot to commit README.md, sorry. Now you could try building it with instructions in README.md

**Comment by masatake** [comment]
I'm sorry for working on this pull request so slowly.
I cannot find a bulk time to concentrate on this topic.

https://github.com/openSUSE/libsolv/tree/master may have all I must study to apply cmake to libreadtags.

I wonder we should prepare FindLibReadtags.cmake like https://github.com/openSUSE/libsolv/blob/master/cmake/modules/FindLibSolv.cmake .

My concerns (and interests):

- [ ] *Unify* the places where the maintainer specifies various versions. 
- [ ] Run the test cases as we do in autotools. (both locally and CI/CD environments)
- [ ] Generate and install libreadtags.pc as we do in autotools.
- [ ] Prepare make install target as we do in autotools.
- [ ] Prepare FindLibSolv.cmake and the way to install it.
- [ ] ...

**Comment by ParticleG** [comment]
> I'm sorry for working on this pull request so slowly. I cannot find a bulk time to concentrate on this topic.
> 
> https://github.com/openSUSE/libsolv/tree/master may have all I must study to apply cmake to libreadtags.
> 
> I wonder we should prepare FindLibReadtags.cmake like https://github.com/openSUSE/libsolv/blob/master/cmake/modules/FindLibSolv.cmake .
> 
> My concerns (and interests):
> 
> * [ ]  _Unify_ the places where the maintainer specifies various versions.
> * [ ]  Run the test cases as we do in autotools. (both locally and CI/CD environments)
> * [ ]  Generate and install libreadtags.pc as we do in autotools.
> * [ ]  Prepare make install target as we do in autotools.
> * [ ]  Prepare FindLibSolv.cmake and the way to install it.
> * [ ]  ...

Actually, if we prepare CMakeList.txt properly, other CMake projects just need to use [find_package](https://cmake.org/cmake/help/latest/command/find_package.html#command:find_package) method to load this package, see [Finding Packages](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Finding%20Packages.html#creating-cmake-package-configuration-files). But if you use CMake method `FetchContent` to add libreadtags to your project, then current changes should be enough. I've nerver tried to create proper "CMake Package Configuration Files", since I'm coding on Windows and use [vcpkg](https://github.com/microsoft/vcpkg) to manage CMake packages. Maybe I'll try that in my spare time.

By the way, OpenSUSE's CMake standard is too old ([2.x.x](https://github.com/openSUSE/libsolv/blob/b3aa55e65f3f6519e7ed8d997d7c708310c8d06e/CMakeLists.txt#L3)) which is deprecated according to CMake official docs. If you want to learn more about CMake, you should look for something more modern (`> 3.5`), like [IXWebSocket](https://github.com/machinezone/IXWebSocket/blob/master/CMakeLists.txt), it also have FindXXX.cmake in it [FindDeflate.cmake](https://github.com/machinezone/IXWebSocket/blob/master/CMake/FindDeflate.cmake)

**Comment by masatake** [comment]
Sorry for making you wait so long.
Looks good. The changes for README.md is excellent. All I wanted to know is written in the changes.

I'll take over this to adjust minor things.
See #53.

**Pull request closed without merging**